Authentication/SignUp/ViewModel/SignupViewModel.swift

```diff
@@ -47,54 +52,85 @@ public final class SignupViewModel<User: UserType, SessionService: SessionServic
     public let passwordValidationErrors: AnyProperty<[String]>
     public let passwordConfirmationValidationErrors: AnyProperty<[String]>
     
+    public let showPassword = MutableProperty(false)
+    public let showConfirmationPassword = MutableProperty(false)
+    
     public var signUpCocoaAction: CocoaAction { return _signUp.unsafeCocoaAction }
     public var signUpErrors: Signal<SessionServiceError, NoError> { return _signUp.errors }
     public var signUpExecuting: Signal<Bool, NoError> { return _signUp.executing.signal }
     
-    private lazy var _signUp: Action<AnyObject, User, SessionServiceError> = {
-        return Action(enabledIf: self._credentialsAreValid) { [unowned self] _ in
-            if let email = Email(raw: self.email.value) {
-                return self._sessionService.signUp(self.name.value, email: email, password: self.password.value)
-            } else {
-                return SignalProducer(error: .InvalidSignUpCredentials(.None)).observeOn(UIScheduler())
-            }
-        }
-    }()
+    public var togglePasswordVisibilityCocoaAction: CocoaAction { return _togglePasswordVisibility.unsafeCocoaAction }
+    public var toggleConfirmPswdVisibilityCocoaAction: CocoaAction { return _toggleConfirmationPasswordVisibility.unsafeCocoaAction }
     
-    internal init(sessionService: SessionService, credentialsValidator: SignupCredentialsValidator = SignupCredentialsValidator()) {
+    private lazy var _signUp: Action<AnyObject, User, SessionServiceError> = self.initializeSignUpAction()
+    
+    private lazy var _togglePasswordVisibility: Action<AnyObject, Bool, NoError> = self.initializeTogglePasswordVisibilityAction()
+    private lazy var _toggleConfirmationPasswordVisibility: Action<AnyObject, Bool, NoError> = self.initializeToggleConfirmationPasswordVisibilityAction()
+    
+    internal init(sessionService: SessionService,
+         credentialsValidator: SignupCredentialsValidator = SignupCredentialsValidator(),
+         passwordConfirmationEnabled: Bool = true,
+         usernameEnabled: Bool = true) {
         _sessionService = sessionService
         
         let nameValidationResult = name.signal.map(credentialsValidator.nameValidator.validate)
         let emailValidationResult = email.signal.map(credentialsValidator.emailValidator.validate)
         let passwordValidationResult = password.signal.map(credentialsValidator.passwordValidator.validate)
         let passwordConfirmValidationResult = combineLatest(password.signal, passwordConfirmation.signal)
```

Comment:
I would avoid naming this Result, one can think that it is talking about Result dependency